static/src/javascripts/projects/common/modules/spacefinder.ts

```diff
@@ -0,0 +1,448 @@
+	minAbove: number;
+	// minimum from (top of) para to bottom of article
+	minBelow: number;
+	// vertical px to clear the content meta element (byline etc) by. 0 to ignore
+	// used for carrot ads
+	clearContentMeta?: number;
+	// custom rules using selectors.
+	selectors?: Record<string, RuleSpacing>;
+	// will run each slot through this fn to check if it must be counted in
+	filter?: (x: SpacefinderItem, lastWinner?: SpacefinderItem) => boolean;
+	// will remove slots before this one
+	startAt?: HTMLElement;
+	// will remove slots from this one on
+	stopAt?: HTMLElement;
+	// will reverse the order of slots (this is useful for lazy loaded content)
+	fromBottom?: boolean;
+};
+
+type SpacefinderWriter = (paras: HTMLElement[]) => Promise<void>;
+
+type SpacefinderOptions = {
+	waitForLinks?: boolean;
+	waitForImages?: boolean;
+	waitForInteractives?: boolean;
+	debug?: boolean;
+};
+
+type ExcludedItem = SpacefinderItem | HTMLElement;
+
+type SpacefinderExclusions = Record<string, ExcludedItem[]>;
+
+type ElementDimensionMap = Record<string, SpacefinderItem[]>;
+
+type Measurements = {
+	bodyTop: number;
+	bodyHeight: number;
+	candidates: SpacefinderItem[];
+	contentMeta?: SpacefinderItem;
+	opponents?: ElementDimensionMap;
+};
+
+const query = (selector: string, context?: HTMLElement | Document) => [
+	...(context ?? document).querySelectorAll<HTMLElement>(selector),
+];
+
+// maximum time (in ms) to wait for images to be loaded and rich links
+// to be upgraded
+const LOADING_TIMEOUT = 5000;
+
+const defaultOptions = {
+	waitForImages: true,
+	waitForLinks: true,
+	waitForInteractives: false,
+	debug: false,
+};
+
+const isIframe = (node: Node): node is HTMLIFrameElement =>
+	node instanceof HTMLIFrameElement;
+
+const isIframeLoaded = (iframe: HTMLIFrameElement) => {
+	// TODO do we need this try / catch?
+	try {
+		return iframe.contentWindow?.document.readyState === 'complete';
+	} catch (err) {
+		return true;
+	}
+};
+
+const getFuncId = (rules: SpacefinderRules) => rules.bodySelector || 'document';
+
+const enableImageLoadingFix = () =>
+	!isInVariantSynchronous(spacefinderOkrMegaTest, 'control');
+
+const onImagesLoaded = enableImageLoadingFix()
+	? onImagesLoadedFixed
+	: onImagesLoadedBroken;
+
+const onRichLinksUpgraded = memoize(
+	(rules: SpacefinderRules) =>
+		query('.element-rich-link--not-upgraded', rules.body).length === 0
+			? Promise.resolve()
+			: new Promise((resolve) => {
+					mediator.once('rich-link:loaded', resolve);
+			  }),
+	getFuncId,
+);
+
+const onInteractivesLoaded = memoize((rules: SpacefinderRules) => {
+	const notLoaded = query('.element-interactive', rules.body).filter(
+		(interactive) => {
+			const iframes = Array.from(interactive.children).filter(isIframe);
+			return !(iframes.length && isIframeLoaded(iframes[0]));
+		},
+	);
+
+	if (notLoaded.length === 0 || !('MutationObserver' in window))
+		return Promise.resolve();
+
+	return Promise.all(
+		notLoaded.map(
```

Comment:
I think this would gain in clarity if you refactored like so:

```ts
// mark the memoized function as async

const mutations = notLoaded.map( /* … */ )
// drop the .then(() => undefined) in favour of
await Promise.all(mutations)
```

Reply:
agreed